### ISSUE TITLE
Reduce CPU usage of batch span processor

### ISSUE THREAD

**Issue by sbandadd** [open-telemetry/opentelemetry-java#2968]
**Is your feature request related to a problem? Please describe.**
Batch span processor is too aggressive for services that want very less tracing overhead. Batch span processor exporter thread constantly polls every 100ms and there is no way to avoid this currently. 

**Describe the solution you'd like**
A short term solution is providing an option for increasing the queue poll wait time. A better solution would be sending a signal to exporter thread when queue size reaches maxExportBatchSize, this way exporter thread is doing no extra work unless it has to. The later solution might not suit better for busy systems who wants very low latency, in that scenario creating a new span processor (using lmax disruptor) and letting users chose among them is better. 

**Describe alternatives you've considered**
None

**Additional context**
Here are the load tests results to reveal the problem.

- The load test is run against a kotlin grpc microservice running in k8s cluster. The service supports createMessage and getMessage grpc API calls which simply read/write from amazon SQS. The service is using OpenTelemetry javaagent for auto instrumentation.
- Allocated 1000m cpu and 1000Mi memory for the service with a single replica.
- Synthetic createMessage/getMessage load is generated from worker pods.
- I am running these tests against the branch https://github.com/open-telemetry/opentelemetry-java/tree/v0.17.0 and using open telemetry java agent for auto instrumentation.
- All the experiments are using the following common parameters when applicable:
  - Otel.bsp.schedule.delay = 300s. This is intentionally set high to reveal the problem.
  - Otel.max.queue.size = 2048.
  - Otel.max.export.batch.size = 5120. This is intentionally set higher than queue size to reveal the problem.
  
In these tests we are interested in understanding the CPU overhead with varying sampler probabilities. Details of the various versions are:
- Otel 0.18.0 disabled: Disabled OpenTelemetry to establish a baseline.
- Otel 0.18.0 enabled with no exporter: Disabled exporter but the service is still using with OpenTelemetry agent.
- [Otel 0.18.0](https://github.com/open-telemetry/opentelemetry-java/pull/2969): This is the custom javaagent built from the brach v0.17.0 supporting the option otel.bsp.queue.poll_wait_time.
- [Otel 0.18.0 with signalling](https://github.com/sbandadd/opentelemetry-java/pull/1): This is a custom javaagent I built using batch signaling where exporter thread is sent a signal when  queue size reaches maxExportBatchSize.
- **Edit** (This is not accurate, I was bitten by linear behavior of size() function for java's inbuilt concurrent queue). Otel 0.18.0 with signalling + concurrent queue: Similar to the above solution except it uses a concurrent queue. I did this based on hope even though in practice concurrent queue has busy spin overheads. 

From the tests we can see that CPU usage stays the same with reduced sampling using the polling. Batched signaling offers better behavior.

![image](https://user-images.githubusercontent.com/62265954/109745684-18b23400-7b89-11eb-99da-c885724b5d85.png)

Here is a screenshot from yjp profiler that revealed me the BatchSpanProcessor thread CPU usage.

![image](https://user-images.githubusercontent.com/62265954/109747045-4dbf8600-7b8b-11eb-8d7e-bb34137f8d3d.png)

The issue seems to be arising from this PR https://github.com/open-telemetry/opentelemetry-java/pull/1579/files which optimized for reduced dropped span rate, but I think one can configure max queue size correctly to reduce the dropped spans.

I have a branch ready with configurable queue poll wait time if someone wants to review.

**Comment by anuraaga** [comment]
I don't see any reason not to allow configuring the poll interval. I don't think we can do `A better solution would be sending a signal to exporter thread when queue size reaches maxExportBatchSize` since we do want to send spans periodically too with default settings even if there aren't many spans, not only based on the max export size, unless I'm missing something about the idea.

I'm not so familiar with YourKit, but that screenshot confuses me - the thread is parked so it shouldn't be counted as any CPU. Just wondering what it means for it to say 36ms used?

**Comment by anuraaga** [comment]
Ah but we can change the poll to just be to the nextExportTime if we allow sending flush requests through the queue, the idea came across at the time but we didn't do it yet for simplicity. Is that the approach your branch takes? Can you share a link?

**Comment by jkwatson** [comment]
Have you checked out the disruptor based span processor that we already have?

**Comment by sbandadd** [comment]
@anuraaga Here is the branch https://github.com/open-telemetry/opentelemetry-java/pull/2969/files for configuring the poll time. I guess using nextExportTime should help as well. Regarding the yjp profiling, it says the thread state is waiting at that instant but consumed 36ms CPU during that **profiling sample**. The cpu cost might be coming from the context switching.

> since we do want to send spans periodically too with default settings.
Yes certainly and I didn't clarify the solution completely. We could simply use scheduleDelayNanos timeout when waiting for signal and proceed to export irrespective of the the queue size.

@jkwatson No I didn't. Could you provide details on how to configure that in auto instrumentation?

**Comment by jkwatson** [comment]
I'm confused by the numbers you've posted. Having a long polling interval only seems helpful in a very narrow use case of extremely low sampling percentage. Can you explain this? Can you provide some jmh based benchmarks that reproduce this?

**Comment by anuraaga** [comment]
@jkwatson I think the idea is that the CPU overhead should be low if there aren't many spans anyways, hence the low sampling percentage. It's true that if we're not doing much work for tracing, it's nice if the BSP also isn't doing much extra work.

**Comment by jkwatson** [comment]
> @jkwatson I think the idea is that the CPU overhead should be low if there aren't many spans anyways, hence the low sampling percentage. It's true that if we're not doing much work for tracing, it's nice if the BSP also isn't doing much extra work.

But, in that case, it seems like it just doesn't matter. A few extra cycles polling for spans seems like it just doesn't matter if we're basically never seen any spans in the first place due to sampling. 

I'm just worried that this is not measuring anything that's actually interesting or useful in the real world, outside of this very narrow benchmark (which we haven't actually seen the source to).

**Comment by jkwatson** [comment]
@sbandadd I don't believe the auto-instrumentation project has this as a built-in option right now. Can you do some benchmarking (and please publish the benchmarking code) in this project, without the auto-instrumentation agent involved, to see if we can get some actual reproduceable results before we start making changes here?

**Comment by sbandadd** [comment]
What sampling % do you recommend? In production systems at scale, we need to use very low sampling anyways. Sorry I couldn't setup jmh to work correctly in my environment and me not being familiar with jmh isn't being helpful.

**Comment by jkwatson** [comment]
So... maybe I'm not understanding something, but why would changing the poll timeout change the CPU utilization? Under normal circumstances (i.e. there are sampled spans being send through the BSP), the poll will immediately return because there will be something in the queue to take. I don't think that changing the poll timeout will change anything at all unless the system is basically idle from a span generation perspective, and even then, it will just cause a few less passes through the polling loop. 

So, I think adjusting the poll timeout will do nothing to a system which is generating spans, and is meaningless to a system that is basically quiescent wrt. span generation. 

I think your benchmark is not measuring what you think it's measuring.

**Comment by sbandadd** [comment]
> the poll will immediately return because there will be something in the queue to take ... it will just cause a few less passes through the polling
The poll does decent amount of work such as taking a lock, waiting on condition that involves context switch. Increasing the timeout is simply making the CPU do less work. It is definitely not meaningless to a system that cares about CPU utilization.

**Comment by jkwatson** [comment]
> > the poll will immediately return because there will be something in the queue to take ... it will just cause a few less passes through the polling
> 
> The poll does decent amount of work such as taking a lock, waiting on condition that involves context switch. Increasing the timeout is simply making the CPU do less work. It is definitely not meaningless to a system that cares about CPU utilization.

How does increasing the timeout make the CPU do less work? If there is something waiting in the queue, the timeout is never used. 

A better answer might be to make the thread drain the queue (up to the batch limit) when the `poll` returns with something in it. That will reduce a little bit of extra work, but I suspect not much. 

Remember that the worker thread here is also converting the span into a SpanData, and calling the exporter, which has to convert the SpanData into its native representation before shipping it off to be exported. The worker thread CPU usage is almost certainly dominated by the actual work, not the queue polling. If you can demonstrate with a concrete reproducable benchmark that the polling timeout actually makes a difference, then by all means let's address it. I just want to make sure that we're not making things more complex based on hunches.

**Comment by breedx-splk** [comment]
FWIW, I ran the existing benchmarks with `../../gradlew -PjmhIncludeSingleClass=BatchSpanProcessorBenchmark jmh` with the original implementation, a version that does a `drainTo()` after the `poll()` return non-null, and the original version with increased poll duration.  They are approximately the same performance and it leads me to conclude that the worker loop isn't a bottleneck.

## Original
![image](https://user-images.githubusercontent.com/75337021/109870787-62e5f480-7c1f-11eb-924b-fa23fc3acafb.png)

## drainTo()
![image](https://user-images.githubusercontent.com/75337021/109870866-74c79780-7c1f-11eb-865d-dbce57d345eb.png)

## poll(5000ms)
![image](https://user-images.githubusercontent.com/75337021/109870912-7f822c80-7c1f-11eb-9adf-c74b753e01f9.png)

**Comment by sbandadd** [comment]
@breedx-splk Is there any extra setup required for jmh benchmarks? I am hitting the following error on main branch

```
org.testcontainers.containers.ContainerLaunchException: Container startup failed
        at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:330)
        at org.testcontainers.containers.GenericContainer.start(GenericContainer.java:311)
        at io.opentelemetry.sdk.trace.SpanPipelineBenchmark$AbstractProcessorBenchmark.setup(SpanPipelineBenchmark.java:69)
        at io.opentelemetry.sdk.trace.jmh_generated.SpanPipelineBenchmark_BatchSpanProcessorBenchmark_measureSpanPipeline_jmhTest._jmh_tryInit_f_batchspanprocessorbenchmark0_G(SpanPipelineBenchmark_BatchSpanProcessorBenchmark_measureSpanPipeline_jmhTest.java:434)
        at io.opentelemetry.sdk.trace.jmh_generated.SpanPipelineBenchmark_BatchSpanProcessorBenchmark_measureSpanPipeline_jmhTest.measureSpanPipeline_Throughput(SpanPipelineBenchmark_BatchSpanProcessorBenchmark_measureSpanPipeline_jmhTest.java:71)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at org.openjdk.jmh.runner.BenchmarkHandler$BenchmarkTask.call(BenchmarkHandler.java:453)
        at org.openjdk.jmh.runner.BenchmarkHandler$BenchmarkTask.call(BenchmarkHandler.java:437)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: org.rnorth.ducttape.RetryCountExceededException: Retry limit hit with exception
        at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:88)
        at org.testcontainers.containers.GenericContainer.doStart(GenericContainer.java:323)
        ... 16 more
Caused by: org.testcontainers.containers.ContainerLaunchException: Could not create/start container
        at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:497)
        at org.testcontainers.containers.GenericContainer.lambda$doStart$0(GenericContainer.java:325)
        at org.rnorth.ducttape.unreliables.Unreliables.retryUntilSuccess(Unreliables.java:81)
        ... 17 more
Caused by: java.lang.IllegalStateException: Container exited with code 1
        at org.testcontainers.containers.GenericContainer.tryStart(GenericContainer.java:469)
        ... 19 more



> Task :sdk:trace:jmh FAILED
Benchmark had encountered error, and fail on error was requested

> Task :sdk:trace:jmhReport
JMH Report generated, please open: file:///Users/sbanda/Projects/sbandadd-forks/opentelemetry-java/sdk/trace/build/reports/jmh/index.html

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':sdk:trace:jmh'.
> A failure occurred while executing me.champeau.gradle.IsolatedRunner
   > Error during execution of benchmarks

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.8.2/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 1m 12s
```

**Comment by jkwatson** [comment]
It looks like running the `SpanPipelineBenchmark` requires docker to be running. If you don't have docker, then probably just run other benchmarks directly by name.

**Comment by amitgud-doordash** [comment]
Similar issue with running other benchmarks. Looks like some gradle/jmh/plugin version issue?

```
./gradlew -PjmhIncludeSingleClass=StatsTraceContextBenchmark :sdk:trace:jmh

> Configure project :all
Inferred project: opentelemetry-java, version: 0.18.0-SNAPSHOT

> Task :proto:downloadProtoArchive
Download https://github.com/open-telemetry/opentelemetry-proto/archive/v0.7.0.zip

> Task :sdk:trace:jmhRunBytecodeGenerator FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':sdk:trace:jmhRunBytecodeGenerator'.
> Could not resolve all files for configuration ':sdk:trace:jmh'.
   > Could not find com.google.guava:guava:.
     Required by:
         project :sdk:trace
   > Could not find io.grpc:grpc-api:.
     Required by:
         project :sdk:trace
         project :sdk:trace > project :proto
   > Could not find io.grpc:grpc-netty-shaded:.
     Required by:
         project :sdk:trace
   > Could not find org.testcontainers:testcontainers:.
     Required by:
         project :sdk:trace
   > Could not find com.google.protobuf:protobuf-java:.
     Required by:
         project :sdk:trace > project :proto
   > Could not find io.grpc:grpc-protobuf:.
     Required by:
         project :sdk:trace > project :proto
   > Could not find io.grpc:grpc-stub:.
     Required by:
         project :sdk:trace > project :proto

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.8.2/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 2s
41 actionable tasks: 25 executed, 16 from cache
```

**Comment by jkwatson** [comment]
make sure you're using jdk 11+ for all your builds.

**Comment by amitgud-doordash** [comment]
```
$ java --version
openjdk 11.0.8 2020-07-14
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.8+10)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.8+10, mixed mode)

$ javac --version
javac 11.0.8
```

**Comment by jkwatson** [comment]
@amitgud-doordash what is `StatsTraceContextBenchmark` that you're trying to run?

**Comment by jkwatson** [comment]
> @amitgud-doordash what is `StatsTraceContextBenchmark` that you're trying to run?

If you pulled that line from the build.gradle.kts, it's not a real thing. Try running with the example that @breedx-splk put into the docs/jmh.md file.

**Comment by sbandadd** [comment]
weird, I have docker 3.1.0 running locally. I could run SpanBenchmark and here is the benchmark result with the change https://github.com/sbandadd/opentelemetry-java/pull/1. We could see throughput improving compared to base version with increasing number of threads. I believe this is due to less signaling from the worker threads and less polling from the exporter thread.

Checking only throughput isn't great here actually. We should be comparing CPU utilization as well, will try to take a stab.


## Original
![image](https://user-images.githubusercontent.com/62265954/109901882-9ab86080-7c4e-11eb-98bf-16191e2c9055.png)

# Batched signal
![image](https://user-images.githubusercontent.com/62265954/109901900-a441c880-7c4e-11eb-8717-a10bc35cc99f.png)

**Comment by sbandadd** [comment]
Ran the SpanBenchmark with 50 iterations to get better signal with [this change](https://github.com/sbandadd/opentelemetry-java/pull/1).

# Baseline
![image](https://user-images.githubusercontent.com/62265954/109921711-280cac80-7c71-11eb-8266-4738083d048b.png)

# With batched signal
![image](https://user-images.githubusercontent.com/62265954/109921778-3ce94000-7c71-11eb-8a13-693a2108e948.png)

Reduced gc time as bonus since the queue is pre allocated (ArrayDequeue).
# Baseline
![image](https://user-images.githubusercontent.com/62265954/109921889-66a26700-7c71-11eb-85b6-4c5d6ed3e606.png)

# With batched signal
![image](https://user-images.githubusercontent.com/62265954/109921976-88035300-7c71-11eb-97d8-8cd430f7dc87.png)


Ran the BatchSpanProcessorDroppedSpansBenchmark with 1/2/5/10 thread configuration and verified drop rate is 0, but the new approach is hitting contention quickly.

# Baseline
![image](https://user-images.githubusercontent.com/62265954/109934870-d3256200-7c81-11eb-9546-ef9ee8067470.png)

# With batched signal
![image](https://user-images.githubusercontent.com/62265954/109934897-d9b3d980-7c81-11eb-8bd7-45383f1af693.png)

**Comment by anuraaga** [comment]
@sbandadd Thanks for investigating this to help with BSP performance, this is good stuff. Do you know what the source of the GC improvement is? You mention preallocated with `ArrayDeque` but we're currently using `ArrayBlockingQueue` which is also preallocated. The "batched signal" code looks like it takes locks much more which could be causing the contention, would you be able to also try comparing with https://github.com/open-telemetry/opentelemetry-java/pull/2971 to see the difference?

**Comment by sbandadd** [comment]
My mistake. You are right on the ArrayBlockingQueue being preallocated. With fewer threads the new approach is better as shown by the throughput of BatchSpanProcessorDroppedSpansBenchmark. I will try to look into your PR and see if we can tweak something.

**Comment by sbandadd** [comment]
@anuraaga You approach does take locks as well right with ArrayBlockingQueue offer() and size()?.


```
    public boolean offer(E e) {
        Objects.requireNonNull(e);
        final ReentrantLock lock = this.lock;
        lock.lock();
        try {
            if (count == items.length)
                return false;
            else {
                enqueue(e);
                return true;
            }
        } finally {
            lock.unlock();
        }
    }
```

**Comment by anuraaga** [comment]
@sbandadd Yup it's definitely not lock-free, but the locked sections seem to be smaller. It may not be a big difference, hence needing to compare with benchmarks :) But if similar, and the JDK can handle the locking for us, it means we'll get future performance optimizations for free if they're able to come up with new patterns so we'd prefer that over manual locking if it works.

**Comment by jkwatson** [comment]
It would also be very useful if someone could compare the BSP to the disruptor-based async span processor with apples-to-apples benchmarks.

**Comment by sbandadd** [comment]
@anuraaga I see bit similar results when using your approach. I tried even better approach using bounded concurrent queue to reduce lock contention under lots of threads. The perf improvement is very clear. Updated version is in here [https://github.com/sbandadd/opentelemetry-java/pull/1/files](https://github.com/sbandadd/opentelemetry-java/pull/1/files)

![image](https://user-images.githubusercontent.com/62265954/110023553-26c79800-7ce2-11eb-88dc-e20507f88c1b.png)

Also note that the notifying exported is clearly expensive. I ran the benchmark with the same version as above expect that it notifies exporter on every new span.

![image](https://user-images.githubusercontent.com/62265954/110023983-a8b7c100-7ce2-11eb-9b8c-527015dc08f9.png)

**Comment by bogdandrutu** [comment]
My 2 cents here, the "embedded" batch span processor that we offer as part of our SDK should probably use native java components which work on Android as well.

For the more advance implementation like `disruptor` based or `org.jctools.queues.MpscArrayQueue` based we can offer sdk-extensions that user MUST be able to configure using the autoconfiguration artifact.

Is that a reasonable state?

**Comment by jkwatson** [comment]
Agree that the built-in components should not have external dependencies. Whether or not other advanced implementations MUST be available in auto-configuration probably depends more on the stability and maturity of them. But, I would like the disruptor to be more easily available (and have apples-to-apples benchmarks done on it vs. the BSP), and definitely be options in autoconfiguration when we think that they are ready for primetime.

**Comment by bogdandrutu** [comment]
@jkwatson the disruptor is based on open-census and that is used in production for more than 3y. So it is pretty safe to assume is fine :)

**Comment by jkwatson** [comment]
> @jkwatson the disruptor is based on open-census and that is used in production for more than 3y. So it is pretty safe to assume is fine :)
> 
> Update:
> Stability ways we need a proper API review of the artifact before marking it stable, and maybe consider using shading for disruptor (which I did not do in open-census and have not heard complains)

Yes. I wasn't trying to suggest that it wasn't solid...just that we should go through the same process as our other components, and make sure we have benchmarks for it, as well. :)

**Comment by sbandadd** [comment]
Here is the result from disruptor benchmark. Disruptor has overhead of signaling and an aggressive sleeping time strategy with [1ms default](https://github.com/open-telemetry/opentelemetry-java/blob/64b9badaa50ef5c71fa03a2d068f5b223caf1c63/sdk-extensions/async-processor/src/main/java/io/opentelemetry/sdk/extension/trace/export/DisruptorAsyncSpanProcessorBuilder.java#L22). I didn't dwell into testing these configuration options. @bogdandrutu It would be ideal if you have some suggestion on ideal configuration for processor based on disruptor


# Disruptor with queue size 2048

![image](https://user-images.githubusercontent.com/62265954/110043592-86cb3800-7cfc-11eb-8888-f068463ee04b.png)

# Concurrent queue with batch signal
![image](https://user-images.githubusercontent.com/62265954/110023553-26c79800-7ce2-11eb-88dc-e20507f88c1b.png)

# Basline
![image](https://user-images.githubusercontent.com/62265954/110023983-a8b7c100-7ce2-11eb-9b8c-527015dc08f9.png)


Here is what I used to run the benchmark

**Comment by anuraaga** [comment]
> For the more advance implementation like disruptor based or org.jctools.queues.MpscArrayQueue based we can offer sdk-extensions that user MUST be able to configure using the autoconfiguration artifact.

@bogdandrutu Should we file a spec issue to define the property for configurable span processors?

**Comment by jkwatson** [comment]
@sbandadd can we close this now?

**Comment by franz1981** [comment]
@anuraaga @jkwatson 

I see that the referenced commit is using a CLQ with a separate size tracker to overcome the size complexity of CLQ: I strongly suggest to use MpscUnboundedXaddArrayQueue on JCTools that offer:
- (order of magnitude) better behaviour then any compare-and-set queue alternative, under contention
- GC free if correctly configured or lower allocation rate the CLQ if not
- O(1)-ish size although I suggest to try different strategies: size is linearish but it would invalidate producer and consumer sequences if queries before any offer, slowing down a streaming consumer


The latter suggest that using a LongAdder or any out atomic counter of JCTools has to be prefer...

Let me know if I should open a new issue/pr to track this.

**Comment by jkwatson** [comment]
Please open a new issue for this, if you feel it is needed. Thanks!
